Announcing Scala.js 1.18.1. (#649)
scala-js/scala-js-website: _posts/news/2025-01-07-announcing-scalajs-1.18.0.md

If you need help with anything related to Scala.js, you may find our community [in `#scala-js` on Discord](https://discord.com/invite/scala) and [on Stack Overflow](https://stackoverflow.com/questions/tagged/scala.js).

Bug reports can be filed [on GitHub](https://github.com/scala-js/scala-js/issues).

## Release notes

If upgrading from Scala.js 0.6.x, make sure to read [the release notes of Scala.js 1.0.0]({{ BASE_PATH }}/news/2020/02/25/announcing-scalajs-1.0.0/) first, as they contain a host of important information, including breaking changes.

This is a **minor** release:

* It is backward binary compatible with all earlier versions in the 1.x series: libraries compiled with 1.0.x through 1.17.x can be used with 1.18.0 without change.
* It is *not* forward binary compatible with 1.17.x: libraries compiled with 1.18.0 cannot be used with 1.17.x or earlier.
* It is *not* entirely backward source compatible: it is not guaranteed that a codebase will compile *as is* when upgrading from 1.17.x (in particular in the presence of `-Xfatal-warnings`).

As a reminder, libraries compiled with 0.6.x cannot be used with Scala.js 1.x; they must be republished with 1.x first.

> ## Changes with compatibility concerns

Reviewer: I think we should mention that we run IR checking after the optimizer (maybe a more vulgarized form) and ask to report and explain how to turn IR checking off in case users get hit by this.

### Enforce stricter rules on the use of `this` in the IR of constructors

Exceptionally, **this is theoretically a binary-breaking change**.
It is possible for previously valid IR (Intermediate Representation, as published in binaries on Maven Central) to be rejected by Scala.js 1.18.0.

We now enforce, during IR checking (by default, only enabled in `fullLink` builds), that constructors abide by stricter rules.
Before calling their `super` constructor (or delegate `this()` constructor), they may not refer to the `this` instance.
The only exception is to *assign* a field, as in `this.foo = bar`.
This restriction is in line with a similar rule in JVM bytecode.

> Unfortunately, our compiler plugin for Scala 2.x was previously generating IR that is invalid according to the new rule in one corner case situation.

Reviewer: ```suggestion
Unfortunately, the Scala.js compiler for Scala 2.x was previously generating IR that is invalid according to the new rule in one corner case situation.
```

If there is a `try..catch` expression inside an argument to the `super` constructor, the compiler may lift it to a separate method.
We compiled that method as an *instance* method, using the `this` value, instead of compiling it as a *static* method.

Here is an example:

{% highlight scala %}
import scala.util.control.Breaks._

class Parent(val i: Int) {
  def this(p: Parent) = this(p.i)
  if (i > 40) throw new IllegalArgumentException
}
class Child extends Parent({

>   val p = try new Parent(42) catch { case _: IllegalArgumentException => break() }

Reviewer: IMO this is very confusing, because the `Parent` class is also used as a value. How about:

```suggestion
  val p = try Integer.parseInt("foo") catch { case _: NumberFormatException => break() }
```

And remove the check in `Parent`?

  p
})

object Test extends App {
  breakable(new Child)
}
{% endhighlight %}

If your application depends on a library that contained such a code shape, you may hit the following IR checking error at link time in Scala.js 1.18.0.

{% highlight none %}
Restricted use of `this` before the super constructor call
{% endhighlight %}

If that is the case, the library needs to be recompiled with Scala.js 1.18.0 or later to be usable again.
Make sure to [tell us on the relevant PR](https://github.com/scala-js/scala-js/pull/5019) if that happens.

In order to mitigate the impact of this change, we conducted an audit of (the latest version of) all Scala.js libraries published to Maven Central.
We did not find any library that violated the new rule.
You may still run into the issue if you depend on a private library.

> ### Drop support for Scala 2.12.2 to 2.12.5 and 2.13.0 to 2.13.2

Reviewer: Did we also drop Java 16 support? (https://github.com/scala-js/scala-js/pull/5073) or did we do this earlier and is this just an artifact?

Reviewer: Ah, OK. Just a misunderstanding on my part then.

Scala 2.12.6+ and 2.13.3+ are still supported.

We still had a number of dedicated code paths, special cases and exceptions for these versions in our codebase.
Moreover, we occasionally hit compiler bugs in early 2.12.x versions, which imposed constraints on how we wrote our code.

### The Wasm backend now requires full support of `try_table` (notably Node.js 23)

If you are using the WebAssembly backend, you now need at least Node.js 23 to run and test.

In Scala.js 1.17.0, the WebAssembly backend worked around some issues in Wasm engines regarding the `try_table` instruction.
Our main blocker was the version of V8 shipped with Node.js 22.
Now that Node.js 23 has been released with the correct behavior, we removed the workaround.

### Changes to the IR and linker APIs

For tooling authors who directly manipulate the IR and linker APIs, there have been some breaking changes in that area.
This is in line with our versioning policy for the linker APIs.

The changes you are most likely to hit are:

* We renamed `NoType` to `VoidType`, which represents the IR type `void`.

>   `void` is now generally handled more regularly, as any other type of the type system.

Reviewer: ```suggestion
  `void` is now generally handled more regularly.
```

?

* `VarRef` now directly contains a `LocalName`, rather than a `LocalIdent` itself containing a `LocalName`.
* `This` is not a separate IR node anymore; instead it is a `VarRef` with the special name `LocalName.This`.
* `ArrayLength`, `Throw`, and a number of more obscure IR nodes have been merged into `UnaryOp`.
* `Transformers` does not make the distinction between "statement" and "expression" anymore.

>   This follows from handling `void` more uniformly.

Reviewer: Just to check: I guess you have deliberately decided against mentioning LinkTimeProperty?

Author: Yes ... but now I reconsidered so I added it.
